navigator_helpers/tasks/sample_to_dataset/task_suite.py

```diff
@@ -0,0 +1,835 @@
+
+        Args:
+            sample_dataset (pd.DataFrame): The sample dataset.
+            system_prompt_type (str, optional): Type of system prompt to use. Defaults to 'cognition'.
+
+        Returns:
+            Tuple[str, dict]: A tuple containing the extracted thinking and data seeds.
+        """
+        data_jsonl = sample_dataset.to_json(orient="records", lines=True)
+        data_schema = str(list(sample_dataset.columns))
+
+        dataseed_prompt = DATASEED_REVERSE_ENG_PROMPT_TEMPLATE.format(
+            sampled_dataset_jsonl=data_jsonl, sampled_dataset_column_list=data_schema
+        )
+        if self.config.verbose:
+            print("------------------- Dataseed prompt ------------")
+            print(dataseed_prompt)
+
+        thinking, data_seeds = self.execute_prompt(dataseed_prompt, system_prompt_type)
+
+        return thinking, data_seeds
+
+    def crowdsource_data_seeds(
+        self,
+        sample_dataset: pd.DataFrame,
+        crowd_size=5,
+        max_num_seeds=4,
+        system_prompt_type="cognition",
+    ) -> dict:
+        """
+        Perform crowd-sourcing by executing the dataseed extraction prompt multiple times,
+        and generating a deduped and ranked list of high-quality data seeds.
+
+        Args:
+            sample_dataset (pd.DataFrame): The sample dataset.
+            crowd_size (int, optional): Number of times to run the user prompt. Defaults to 5.
+            max_num_seeds (int, optional): Maximum number of seeds to return. Defaults to 4.
+            system_prompt_type (str, optional): Type of system prompt to use. Defaults to 'cognition'.
+
+        Returns:
+            dict: A dictionary containing the ranked data seeds.
+
+        Raises:
+            ValueError: If all attempts to extract data seeds fail.
+            RuntimeError: If ranking fails after multiple attempts.
+        """
+
+        # Collect data seeds from crowd_size runs of execute_prompt_w_cognition
+        data_seeds_list = []
+        self.logger.info(f"  |-- Crowd size: {crowd_size}")
+        for i in range(crowd_size):
+            self.logger.info(f"  |-- 🫡 assistant opinion {i+1}")
+            try:
+                _, data_seeds = self.extract_data_seeds_in_one_shot(
+                    sample_dataset, system_prompt_type
+                )
+
+                # Define the model for individual column entries
+                class ExampleDataSeedColumnModel(BaseModel):
+                    column_name: str
+                    description: str
+                    example_values: List[Union[str, int, bool]]
+
+                # Define the model for the entire structure
+                class ExampleDataSeedModel(BaseModel):
+                    columns: List[ExampleDataSeedColumnModel]
+
+                is_valid_data_seeds, validation_result = validate_json_with_pydantic(
+                    ExampleDataSeedModel, data_seeds
+                )
```

Comment:
I'm a bit allergic to the model definitions living inside a class method like this, but your validation with pydantic trick is awesome 👏

Reply:
+1, I think in this case these class definitions could be defined at the top level? They don't change for each iteration of this for, so I'd say moving them to the top level would be a bit clear (I would expect that if they are defined within the for loop, they would use some local variables, but they don't in this case).

Reply:
Addressed  in https://github.com/gretelai/navigator-helpers/pull/34/commits/59fb69f5ced87736cd14ae5358def58fb9ca7635